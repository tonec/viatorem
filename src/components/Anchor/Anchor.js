import React, { Component } from 'react'
import PropTypes from 'prop-types'
import createChainedFunction from 'utils/createChainedFunction'

class Anchor extends Component {

  static propTypes = {
    href: PropTypes.string,
    onClick: PropTypes.func,
    onKeyDown: PropTypes.func,
    disabled: PropTypes.bool,
    tabIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    className: PropTypes.string,
    children: PropTypes.node.isRequired
  }

  static defaultProps = {
    href: '',
    onClick: null,
    onKeyDown: null,
    disabled: false,
    tabIndex: 0,
    className: ''
  }

  isTrivialHref (href) {
    return !href || href.trim() === '#'
  }

  handleClick = event => {
    const { disabled, href, onClick } = this.props

    if (disabled || this.isTrivialHref(href)) {
      event.preventDefault()
    }

    if (disabled) {
      event.stopPropagation()
      return
    }

    if (onClick) {
      onClick(event)
    }
  }

  handleKeyDown = event => {
    if (event.key === ' ') {
      event.preventDefault()
      this.handleClick(event)
    }
  }

  render () {
    const { tabIndex, disabled, onKeyDown, children, ...props } = this.props

    if (this.isTrivialHref(props.href)) {
      props.href = props.href || '#'
    }

    if (disabled) {
      props.tabIndex = -1
      props.style = { pointerEvents: 'none', ...props.style }
    }

    return (
      <a
        {...props}
        onClick={this.handleClick}
        onKeyDown={createChainedFunction(this.handleKeyDown, onKeyDown)}
      >
        {children}
      </a>
    )
  }
}

export default Anchor
