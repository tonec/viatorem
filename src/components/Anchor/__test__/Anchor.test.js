import React from 'react'
import { shallow } from 'enzyme'
import Anchor from '../Anchor'

describe('Anchor (snapshot)', () => {

  const props = {
    href: 'http://www.example.com',
    onClick: () => {},
    disabled: true,
    className: 'test-class-name'
  }

  it('renders according to snapshot', () => {
    const component = shallow(<Anchor {...props}>Link text</Anchor>)
    expect(component).toMatchSnapshot()
  })
})
