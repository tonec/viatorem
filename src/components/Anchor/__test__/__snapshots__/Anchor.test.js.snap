// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anchor (snapshot) renders according to snapshot 1`] = `
<a
  className="test-class-name"
  href="http://www.example.com"
  onClick={[Function]}
  onKeyDown={[Function]}
  style={
    Object {
      "pointerEvents": "none",
    }
  }
  tabIndex={-1}
>
  Link text
</a>
`;
